### PR TITLE
Get multiple properties

### PR DESCRIPTION
**Pull request by rickardoberg**
This adds support for retrieving multiple properties at a time from nodes/rels, and updates various client code that previously used getPropertyKeys and looping to use this new method instead.

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java [hunk 1]
```diff
@@ -377,6 +382,85 @@ public Object getProperty( String key, Object defaultValue )
     }
 
     @Override
+    public Map<String, Object> getProperties( String... keys )
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java]: We can add a check `keys == null || keys.length == 0` and do a quick return of `Collections.emptyMap()`

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java [hunk 2]
```diff
@@ -377,6 +382,85 @@ public Object getProperty( String key, Object defaultValue )
...
+                    }
+
+                    while ( propertyCursor.next() )
+                    {
+                        for ( int i = 0; i < propertyKeys.length; i++ )
+                        {
+                            int propertyKey = propertyKeys[i];
+                            if ( propertyKey == propertyCursor.get().propertyKeyId() )
+                            {
+                                properties.put( keys[i], propertyCursor.get().value() );
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java]: how about `break;` when we have found the matching key?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java [hunk 3]
```diff
@@ -236,6 +241,85 @@ public RelationshipType getType()
...
+        {
+            try ( Cursor<RelationshipItem> relationship = statement.readOperations().relationshipCursor( getId() ) )
+            {
+                if ( !relationship.next() )
+                {
+                    throw new NotFoundException( "Relationship not found",
+                            new EntityNotFoundException( EntityType.RELATIONSHIP, getId() ) );
+                }
+
+                try ( Cursor<PropertyItem> propertyCursor = relationship.get().properties() )
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java:269]: Can we create a utility-class that contains a utility method for this code that is duplicated between NodeProxy and RelationshipProxy?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java (hunk 1 above)
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java]: `keys` should be `names`

File: community/kernel/src/test/java/org/neo4j/kernel/impl/core/NodeTest.java [hunk 4]
```diff
@@ -326,9 +327,14 @@ public void testNodeGetProperties()
...
         keys.next();
         keys.next();
-        assertTrue( node1.hasProperty( key1 ) );
-        assertTrue( node1.hasProperty( key2 ) );
-        assertTrue( node1.hasProperty( key3 ) );
+        Map<String, Object> properties = node1.getAllProperties();
+        assertTrue( properties.get( key1 ).equals( int1 ) );
+        assertTrue( properties.get( key2 ).equals( int2 ) );
+        assertTrue( properties.get( key3 ).equals( string ) );
+        properties = node1.getProperties( key1, key2 );
```
Comment (lutovich) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/core/NodeTest.java:334]: Missing tests for negative cases where given array of keys is empty or null

File: community/kernel/src/test/java/org/neo4j/kernel/impl/core/TestRelationship.java [hunk 5]
```diff
@@ -557,6 +559,16 @@ public void testRelGetProperties()
         assertTrue( rel1.hasProperty( key1 ) );
         assertTrue( rel1.hasProperty( key2 ) );
         assertTrue( rel1.hasProperty( key3 ) );
+
+        Map<String, Object> properties = rel1.getAllProperties();
+        assertTrue( properties.get( key1 ).equals( int1 ) );
+        assertTrue( properties.get( key2 ).equals( int2 ) );
+        assertTrue( properties.get( key3 ).equals( string ) );
+        properties = rel1.getProperties( key1, key2 );
```
Comment (lutovich) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/core/TestRelationship.java:567]: Missing tests for negative cases where given array of keys is empty or null

File: community/kernel/src/main/java/org/neo4j/graphdb/PropertyContainer.java [hunk 6]
```diff
@@ -126,4 +128,19 @@
      */
     // TODO: figure out concurrency semantics
     Iterable<String> getPropertyKeys();
+
+    /**
+     * Returns specified existing properties.
+     *
+     * @param keys the property keys to return
+     * @return specified properties on this property container
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/graphdb/PropertyContainer.java]: We should ask users to not rely on the mutability of the returned map and document what will happen if empty/null array of names is given as a parameter

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java [hunk 7]
```diff
@@ -377,6 +382,85 @@ public Object getProperty( String key, Object defaultValue )
...
+                            }
```
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java]: Would it be more expensive to have a more dynamic structure for `propertyKeys` and removing the keys that we already found to avoid looping over them again?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java [hunk 8]
```diff
@@ -377,6 +382,85 @@ public Object getProperty( String key, Object defaultValue )
...
+                {
+                    Map<String, Object> properties = new HashMap<>( ((int) (keys.length * 1.3f)) );
+
+                    // Specific properties given
+                    int[] propertyKeys = new int[keys.length];
+                    for ( int i = 0; i < keys.length; i++ )
+                    {
+                        String key = keys[i];
+                        propertyKeys[i] = statement.readOperations().propertyKeyGetForName( key );
+                    }
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java]: there is no reason to create an array here

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/GraphPropertiesProxy.java [hunk 9]
```diff
@@ -195,6 +198,45 @@ public Object removeProperty( String key )
...
+    public Map<String, Object> getProperties( String... names )
+    {
+        try ( Statement statement = actions.statement() )
+        {
+            Map<String, Object> properties = new HashMap<>();
+            ReadOperations readOperations = statement.readOperations();
+            for ( String name : names )
+            {
+                int propertyKeyId = readOperations.propertyKeyGetForName( name );
+                properties.put( name, readOperations.graphGetProperty( propertyKeyId ) );
```
Comment (AlexanderYastrebov) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/GraphPropertiesProxy.java]: Here you could end up with null values in map for absent properties. Is that ok?

Comment (thobe) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/GraphPropertiesProxy.java]: That might actually be desirable (since it means the set of keys is maintained), I wonder if we should make the other implementations do that as well, or if we should change this implementation, they should at least have the same behaviour in that regard (although this implementation is of course never used by end-user code).

This is something we should discuss IMO.

File: community/kernel/src/main/java/org/neo4j/graphdb/PropertyContainer.java (hunk 6 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/graphdb/PropertyContainer.java]: I hope we make sure that the returned map is immutable.

What about saying that we'll throw some exception (NPE or IAE) if we get `null` keys in the array?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/GraphPropertiesProxy.java [hunk 10]
```diff
@@ -195,6 +198,45 @@ public Object removeProperty( String key )
...
+            }
+            return properties;
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/GraphPropertiesProxy.java]: Please return `Collections.unmodifiableMap( properties )` instead.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/GraphPropertiesProxy.java [hunk 11]
```diff
@@ -195,6 +198,45 @@ public Object removeProperty( String key )
...
+            Map<String, Object> properties = new HashMap<>();
+            ReadOperations readOperations = statement.readOperations();
+            PrimitiveIntIterator propertyKeys = readOperations.graphGetPropertyKeys();
+            while ( propertyKeys.hasNext() )
+            {
+                int propertyKeyId = propertyKeys.next();
+                properties.put( readOperations.propertyKeyGetName( propertyKeyId ),
+                        readOperations.graphGetProperty( propertyKeyId ) );
+            }
+            return properties;
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/GraphPropertiesProxy.java:234]: Please return `Collections.unmodifiableMap( properties )` instead.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/GraphPropertiesProxy.java (hunk 9 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/GraphPropertiesProxy.java]: +1 on the discussion.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java (hunk 1 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java]: I think we should throw if `keys == null` or if `keys` contains `null` values.
We can return `emptyMap()` when `keys.length == 0`.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java [hunk 12]
```diff
@@ -377,6 +382,85 @@ public Object getProperty( String key, Object defaultValue )
...
+                        {
+                            int propertyKey = propertyKeys[i];
+                            if ( propertyKey == propertyCursor.get().propertyKeyId() )
+                            {
+                                properties.put( keys[i], propertyCursor.get().value() );
+                            }
+                        }
+                    }
+
+                    return properties;
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java]: Please return `Collections.unmodifiableMap( properties )` instead.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java [hunk 13]
```diff
@@ -377,6 +382,85 @@ public Object getProperty( String key, Object defaultValue )
...
+
+                    // Get all properties
+                    while ( propertyCursor.next() )
+                    {
+                        String name = statement.readOperations().propertyKeyGetName(
+                                propertyCursor.get().propertyKeyId() );
+                        properties.put( name, propertyCursor.get().value() );
+                    }
+
+                    return properties;
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java:443]: Please return `Collections.unmodifiableMap( properties )` instead.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java [hunk 14]
```diff
@@ -377,6 +382,85 @@ public Object getProperty( String key, Object defaultValue )
...
+            {
+                if ( !node.next() )
+                {
+                    throw new NotFoundException( "Node not found",
+                            new EntityNotFoundException( EntityType.NODE, getId() ) );
+                }
+
+                try ( Cursor<PropertyItem> propertyCursor = node.get().properties() )
+                {
+                    Map<String, Object> properties = new HashMap<>( ((int) (keys.length * 1.3f)) );
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java]: Can we add a comment explaining the rationale behind `((int) (keys.length * 1.3f))`?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java [hunk 15]
```diff
@@ -236,6 +241,85 @@ public RelationshipType getType()
...
+                {
+                    Map<String, Object> properties = new HashMap<>( ((int) (keys.length * 1.3f)) );
+
+                    // Specific properties given
+                    int[] propertyKeys = new int[keys.length];
+                    for ( int i = 0; i < keys.length; i++ )
+                    {
+                        String key = keys[i];
+                        propertyKeys[i] = statement.readOperations().propertyKeyGetForName( key );
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java]: no reason to create an array here

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java [hunk 16]
```diff
@@ -236,6 +241,85 @@ public RelationshipType getType()
...
+                        {
+                            int propertyKey = propertyKeys[i];
+                            if ( propertyKey == propertyCursor.get().propertyKeyId() )
+                            {
+                                properties.put( keys[i], propertyCursor.get().value() );
+                            }
+                        }
+                    }
+
+                    return properties;
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java]: Please return `Collections.unmodifiableMap( properties )` instead.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java [hunk 17]
```diff
@@ -236,6 +241,85 @@ public RelationshipType getType()
...
+
+                    // Get all properties
+                    while ( propertyCursor.next() )
+                    {
+                        String name = statement.readOperations().propertyKeyGetName(
+                                propertyCursor.get().propertyKeyId() );
+                        properties.put( name, propertyCursor.get().value() );
+                    }
+
+                    return properties;
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java:302]: Please return `Collections.unmodifiableMap( properties )` instead.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java (hunk 3 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java:269]: +1

File: community/kernel/src/test/java/org/neo4j/kernel/impl/core/NodeTest.java (hunk 4 above)
Comment (davidegrohmann) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/core/NodeTest.java:334]: I would also like a test when the array contains `null` values.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/core/TestRelationship.java (hunk 5 above)
Comment (davidegrohmann) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/core/TestRelationship.java:567]: I would also like a test when the array contains `null` values.

File: community/kernel/src/test/java/org/neo4j/test/mocking/Properties.java [hunk 18]
```diff
@@ -82,4 +82,9 @@ public Object answer( InvocationOnMock invocation ) throws Throwable
     {
         return properties.keySet().iterator();
     }
+
+    public SortedMap<String, Object> getProperties()
+    {
+        return properties;
```
Comment (davidegrohmann) [inline community/kernel/src/test/java/org/neo4j/test/mocking/Properties.java:88]: Not super important, but why not `Collections.unmodifiableMap( properties )`?

File: community/ndp/messaging-v1/src/main/java/org/neo4j/ndp/messaging/v1/infrastructure/ValueRelationship.java [hunk 19]
```diff
@@ -107,6 +108,26 @@ public Node getEndNode()
...
+        return selected;
+    }
+
+    @Override
+    public Map<String, Object> getAllProperties()
+    {
+        return props;
+    }
+
+    @Override
```
Comment (davidegrohmann) [inline community/ndp/messaging-v1/src/main/java/org/neo4j/ndp/messaging/v1/infrastructure/ValueRelationship.java]: Can we avoid this duplication of code in `NodeValue` and `RelationshipValue`?

File: community/ndp/messaging-v1/src/main/java/org/neo4j/ndp/messaging/v1/infrastructure/ValueUnboundRelationship.java [hunk 20]
```diff
@@ -95,6 +93,26 @@ public long getId()
...
+        return selectedProperties;
+    }
+
+    @Override
+    public Map<String, Object> getAllProperties()
+    {
+        return props;
+    }
+
+    @Override
```
Comment (davidegrohmann) [inline community/ndp/messaging-v1/src/main/java/org/neo4j/ndp/messaging/v1/infrastructure/ValueUnboundRelationship.java]: This code has been duplicate 3 times, please let's try to avoid so much duplication.

File: community/shell/src/main/java/org/neo4j/shell/kernel/ReadOnlyGraphDatabaseProxy.java [hunk 21]
```diff
@@ -387,6 +387,18 @@ public Object getProperty( String key, Object defaultValue )
         }
 
         @Override
+        public Map<String, Object> getProperties( String... names )
+        {
+            return actual.getProperties( names );
```
Comment (davidegrohmann) [inline community/shell/src/main/java/org/neo4j/shell/kernel/ReadOnlyGraphDatabaseProxy.java]: `Collections.unmodifiableMap( ... )`

File: community/shell/src/main/java/org/neo4j/shell/kernel/ReadOnlyGraphDatabaseProxy.java [hunk 22]
```diff
@@ -387,6 +387,18 @@ public Object getProperty( String key, Object defaultValue )
...
+        }
+
+        @Override
+        public Map<String, Object> getAllProperties()
+        {
+            return actual.getAllProperties();
```
Comment (davidegrohmann) [inline community/shell/src/main/java/org/neo4j/shell/kernel/ReadOnlyGraphDatabaseProxy.java:398]: `Collections.unmodifiableMap( ... )`

File: community/shell/src/main/java/org/neo4j/shell/kernel/ReadOnlyGraphDatabaseProxy.java [hunk 23]
```diff
@@ -536,6 +548,18 @@ public Object getProperty( String key, Object defaultValue )
         }
 
         @Override
+        public Map<String, Object> getProperties( String... names )
+        {
+            return actual.getProperties( names );
```
Comment (davidegrohmann) [inline community/shell/src/main/java/org/neo4j/shell/kernel/ReadOnlyGraphDatabaseProxy.java]: `Collections.unmodifiableMap( ... )`

File: community/shell/src/main/java/org/neo4j/shell/kernel/ReadOnlyGraphDatabaseProxy.java [hunk 24]
```diff
@@ -536,6 +548,18 @@ public Object getProperty( String key, Object defaultValue )
...
+        }
+
+        @Override
+        public Map<String, Object> getAllProperties()
+        {
+            return actual.getAllProperties();
```
Comment (davidegrohmann) [inline community/shell/src/main/java/org/neo4j/shell/kernel/ReadOnlyGraphDatabaseProxy.java:398]: `Collections.unmodifiableMap( ... )`

File: community/kernel/src/test/java/org/neo4j/kernel/impl/core/TestRelationship.java [hunk 25]
```diff
@@ -557,6 +559,16 @@ public void testRelGetProperties()
...
+        assertTrue( properties.get( key1 ).equals( int1 ) );
+        assertTrue( properties.get( key2 ).equals( int2 ) );
+        assertFalse( properties.containsKey( key3 ) );
+
```
Comment (davidegrohmann) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/core/TestRelationship.java:571]: Can we please split this method in several ones with proper test-explanatory names?

File: community/kernel/src/test/java/org/neo4j/kernel/impl/core/NodeTest.java [hunk 26]
```diff
@@ -326,9 +327,14 @@ public void testNodeGetProperties()
...
+        assertTrue( properties.get( key1 ).equals( int1 ) );
+        assertTrue( properties.get( key2 ).equals( int2 ) );
+        assertFalse( properties.containsKey( key3 ) );
```
Comment (davidegrohmann) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/core/NodeTest.java:337]: Can we please split this method in several ones with proper test-explanatory names?

**Comment by davidegrohmann** [comment]
Missing unit tests on `NodeValue`, `ValueRelationship`, `ValueUnboundRelationship`, `NodeProxy`, and `RelationshipProxy`?

File: community/kernel/src/test/java/org/neo4j/test/mocking/Properties.java (hunk 18 above)
Comment (rickardoberg) [inline community/kernel/src/test/java/org/neo4j/test/mocking/Properties.java:88]: Two reasons: first I want to minimize the impact of GC that these methods have. Second, if the client changes these maps nothing bad happens. So returning modifiable maps should be fine.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java (hunk 1 above)
Comment (rickardoberg) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java]: @burqen no, we use the word "keys" in the API (e.g. getPropertyKeys), so the correct word is indeed keys

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java (hunk 3 above)
Comment (rickardoberg) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/RelationshipProxy.java:269]: I'll put it in an abstract super class for NodeProxy/RelationshipProxy

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java (hunk 7 above)
Comment (rickardoberg) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java]: Yes, I'm trying to avoid generating unnecessary GC.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java (hunk 8 above)
Comment (rickardoberg) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java]: So you think it would be better to do the name->int lookup on every check instead? Won't that impact performance a bit too much?

File: community/ndp/messaging-v1/src/main/java/org/neo4j/ndp/messaging/v1/infrastructure/ValueUnboundRelationship.java (hunk 20 above)
Comment (rickardoberg) [inline community/ndp/messaging-v1/src/main/java/org/neo4j/ndp/messaging/v1/infrastructure/ValueUnboundRelationship.java]: I'll introduce a common base class

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/GraphPropertiesProxy.java (hunk 9 above)
Comment (rickardoberg) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/GraphPropertiesProxy.java]: I would prefer if only found properties were included, in all cases. That would match better the semantics that we already have in place, i.e. there cannot be null valued properties.

**Comment by lutovich** [comment]
I agree with @davidegrohmann on missing tests and needed refactoring in `TestRelationship` and `NodeTest`

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/PropertyContainerProxy.java [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+ * Base class for NodeProxy/RelationshipProxy with common methods for PropertyContainer
+ */
+public abstract class PropertyContainerProxy
+{
+    protected Map<String, Object> getProperties( Statement statement,
+            Cursor<PropertyItem> propertyCursor,
+            String[] keys )
+    {
+        // Create a map which is slightly larger than the list of keys
+        Map<String, Object> properties = new HashMap<>( ((int) (keys.length * 1.3f)) );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/PropertyContainerProxy.java:39]: can we extract creation of map that should not resize anymore to map util?

Comment (rickardoberg) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/PropertyContainerProxy.java:39]: Why?

Comment (jexp) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/PropertyContainerProxy.java:39]: Actually we should not use an HashMap but a fixed-size, efficient ArrayMap :)

Comment (rickardoberg) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/PropertyContainerProxy.java:39]: No we shouldn't, since ArrayMap does not implement Map. And we're NOT going to expose ArrayMap from our API. Other suggestions?

Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/PropertyContainerProxy.java:39]: because that sounds like handy things to do if you want to avoid useless resizing. 
But will agree with @jexp suggestion :)

Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/PropertyContainerProxy.java:39]: Then falling back to my original comment, that sounds handy to have helper method that will create map with the size that you exactly need.

Comment (thobe) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/PropertyContainerProxy.java:39]: Last time I checked `org.neo4j.kernel.impl.util.ArrayMap` was all but "efficient", and definitely not "fixed-size". We might want to write some new Map implementation for this purpose, but we definitely would not want to reuse `org.neo4j.kernel.impl.util.ArrayMap` here. I furthermore agree with @rickardoberg that the cost of writing our own Map implementation does not seem worth it for this use case. Even if the cost of this conversation is starting to creep up on being as high as writing our own Map implementation would be.

My vote is thus, let the bike shed remain [dark orchid](http://darkorchid.bikeshed.org/).
